lgc/builder/BuilderImpl.cpp

```diff
@@ -688,109 +741,95 @@ Instruction *BuilderImpl::createWaterfallLoop(Instruction *nonUniformInst, Array
+
+  Value *readFirstLane = nullptr;
+
+  // Emit waterfall begin intrinsics. The first begin contains a null token for the previous token argument
+  Value *waterfallBegin = ConstantInt::get(getInt32Ty(), 0);
+  for (auto nonUniformIndex : nonUniformIndices) {
+    if (scalarizeDescriptorLoads && sharedIndex) {
+      // If all the operands share the same non-index, then we emit just one read first lane intrinsic.
+      auto sharedIndexTy = sharedIndex->getType();
+      waterfallBegin = CreateIntrinsic(Intrinsic::amdgcn_waterfall_begin, sharedIndexTy, {waterfallBegin, sharedIndex},
+                                       nullptr, instName);
+      readFirstLane = CreateIntrinsic(Intrinsic::amdgcn_waterfall_readfirstlane, {sharedIndexTy, sharedIndexTy},
+                                      {waterfallBegin, sharedIndex}, nullptr, instName);
+      break;
     }
-
-    // Ensure we do not create a waterfall across blocks.
-    // FIXME: we could use dominator check to allow scalarizing descriptor loads on multi-block spans;
-    // however, this also requires backend support for multi-block waterfalls to be implemented.
-    if (!identicalIndexes || !firstIndexInst ||
-        (firstIndexInst && firstIndexInst->getParent() != nonUniformInst->getParent()))
-      scalarizeDescriptorLoads = false;
+    Value *nonUniformIndex32Bit =
+        nonUniformIndex->getType()->isIntegerTy(64) ? nonUniformIndex32BitVal[nonUniformIndex] : nonUniformIndex;
+    waterfallBegin = CreateIntrinsic(Intrinsic::amdgcn_waterfall_begin, nonUniformIndex32Bit->getType(),
+                                     {waterfallBegin, nonUniformIndex32Bit}, nullptr, instName);
   }
 
-  // Save Builder's insert point
-  IRBuilder<>::InsertPointGuard guard(*this);
-
-  Value *waterfallBegin;
-  if (scalarizeDescriptorLoads) {
-    SetInsertPoint(nonUniformInst);
-    auto descTy = firstIndexInst->getType();
-    // Create waterfall.begin and waterfall.readfirstlane intrinsics.
-    waterfallBegin = ConstantInt::get(getInt32Ty(), 0);
-    waterfallBegin =
-        CreateIntrinsic(Intrinsic::amdgcn_waterfall_begin, descTy, {waterfallBegin, firstIndexInst}, nullptr, instName);
-
-    // Scalarize shared index.
-    Value *readFirstLane = CreateIntrinsic(Intrinsic::amdgcn_waterfall_readfirstlane, {descTy, descTy},
-                                           {waterfallBegin, firstIndexInst}, nullptr, instName);
-
-    for (auto *nonUniformVal : nonUniformIndices) {
-      // Get the first non uniform instruction of the chain.
-      auto it1 = newOrigNonUniformVal.find(nonUniformVal);
-      Value *origNonUniformVal = nonUniformVal;
-      if (it1 != newOrigNonUniformVal.end())
-        origNonUniformVal = it1->second;
-
-      auto [nonUniformImageCallOperand, operandIdx] = nonUniformIndexImageCallOperand[origNonUniformVal];
-
-      if (origNonUniformVal == nonUniformImageCallOperand)
-        continue;
+  for (unsigned operandIdx : operandIdxs) {
```

Comment:
Why does it need to loop through `operandIdxs` rather than `nonUniformIndices`?

Reply:
The amdgcn.waterfall.begin intrinsic has the nonUniformIndex as a parameter. The amdgcn.waterfall.readfirstlane has two parameters: 
1. the first argument is the amdgcn.waterfall.begin intrinsic
2. in case of scalarization, the second argument is the nonUniformIndex. But, if there is not a nonUniformIndex or if the scalrarization is disabled, then the second argument is the operand of the non-uniform instruction. The later is calculated with the help operandIdxs.